zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+                    if asset == asset_in {
+                        *balance = balance.checked_add(&amount_in).ok_or(Error::<T>::MathError)?;
+                    }
+                    *balance = balance.checked_sub(&amount_out).ok_or(Error::<T>::MathError)?;
+                }
+                let new_price = pool.calculate_spot_price(asset_in)?;
+                ensure!(
+                    new_price >= MIN_SPOT_PRICE.saturated_into(),
+                    Error::<T>::SpotPriceBelowMin
+                );
+                Self::deposit_event(Event::<T>::SellExecuted {
+                    who: who.clone(),
+                    market_id,
+                    asset_in,
+                    amount_in,
+                    amount_out,
+                    swap_fee_amount,
+                    external_fee_amount,
+                });
+                Ok(())
+            })
+        }
+
+        #[require_transactional]
+        fn do_join(
+            who: T::AccountId,
+            market_id: MarketIdOf<T>,
+            pool_shares_amount: BalanceOf<T>,
+            max_amounts_in: Vec<BalanceOf<T>>,
+        ) -> DispatchResult {
+            ensure!(pool_shares_amount != Zero::zero(), Error::<T>::ZeroAmount);
+            let market = T::MarketCommons::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Active, Error::<T>::MarketNotActive);
+            Self::try_mutate_pool(&market_id, |pool| {
+                // FIXME Round up to avoid exploits.
+                let ratio = bdiv(
+                    pool_shares_amount.saturated_into(),
+                    pool.liquidity_shares_manager.total_shares()?.saturated_into(),
+                )?;
+                let mut amounts_in = vec![];
+                for (&asset, &max_amount_in) in pool.assets().iter().zip(max_amounts_in.iter()) {
+                    let balance_in_pool = pool.reserve_of(&asset)?;
+                    // FIXME Round up to avoid exploits.
+                    let amount_in = bmul(ratio, balance_in_pool.saturated_into())?.saturated_into();
+                    amounts_in.push(amount_in);
+                    ensure!(amount_in <= max_amount_in, Error::<T>::AmountInAboveMax);
+                    T::AssetManager::transfer(asset, &who, &pool.account_id, amount_in)?;
+                }
+                for ((_, balance), &amount_in) in pool.reserves.iter_mut().zip(amounts_in.iter()) {
+                    *balance = balance.saturating_add(amount_in);
+                }
+                pool.liquidity_shares_manager.join(&who, pool_shares_amount)?;
+                let new_liquidity_parameter = pool.liquidity_parameter.saturating_add(
+                    bmul(ratio.saturated_into(), pool.liquidity_parameter.saturated_into())?
+                        .saturated_into(),
+                );
+                pool.liquidity_parameter = new_liquidity_parameter;
+                Pools::<T>::insert(market_id, pool);
```

Comment:
Isn't this already covered by the surrounding `try_mutate_pool` call?

```suggestion
```

Reply:
9e02d068990fe8baf5f1b0525fff7188d363619d